### PR TITLE
Add Support for US Government cloud deployment

### LINKED ISSUE #3128: Enable deployment to Azure Government Cloud
Deployment fails immediately when attempting to deploy to Azure Gov regions because of the distinct endpoints hardcoded into TRE.  Requesting assisting in removing hardcoded endpoints and allowing endpoints to be determined based on region.  Anywhere the az libraries are used (so long as it is up to date) will use correct endpoints.

### PR DESCRIPTION
**Pull request by LizaShak**
# Resolves #3128 

## What is being addressed

Enable deployment on US Government cloud. 

## How is this addressed

- Use a separate module to get privatelinks
- Remove hardcoded endpoints
- Pass additional parameters on clients
- Specify parameters needed to determine the destination cloud

### CODE REVIEW

File: airlock_processor/shared_code/cloud.py
Comment (martinpeck) [reply airlock_processor/shared_code/cloud.py:16]: i.e. repeated in api_app/core/cloud.py

File: cli/tre/authentication.py
Comment (martinpeck) [inline cli/tre/authentication.py:48]: nit: it's generally better to use `urllib` to parse and manipulate URLs

Comment (martinpeck) [reply cli/tre/authentication.py:48]: also, this is repeated code
perhaps we need a TRE utils package that we can import into multiple things

File: core/terraform/locals.tf
Comment (martinpeck) [inline core/terraform/locals.tf:38]: image

Comment (martinpeck) [reply core/terraform/locals.tf:38]: is there any way we can pad this with comments so it's more obvious what it's expecting, the things it captures, and what it outputs?

File: templates/shared_services/airlock_notifier/terraform/locals.tf
Comment (martinpeck) [inline templates/shared_services/airlock_notifier/terraform/locals.tf:12]: How does the TRE ID get added?
Or...does it?

Comment (martinpeck) [reply templates/shared_services/airlock_notifier/terraform/locals.tf:12]: Ah....we're using the FQDN, that already has it, right?

File: templates/shared_services/cyclecloud/terraform/variables.tf
Comment (martinpeck) [inline templates/shared_services/cyclecloud/terraform/variables.tf:3]: nit: add new-line to end of file

File: templates/shared_services/databricks-auth/terraform/variables.tf
Comment (martinpeck) [inline templates/shared_services/databricks-auth/terraform/variables.tf:11]: nit: add new-line to end of file

File: templates/shared_services/gitea/terraform/variables.tf
Comment (martinpeck) [inline templates/shared_services/gitea/terraform/variables.tf:33]: nit: add new-line to end of file

File: templates/workspace_services/databricks/terraform/variables.tf
Comment (martinpeck) [inline templates/workspace_services/databricks/terraform/variables.tf:26]: nit: add new-line to end of file

File: templates/workspace_services/health-services/terraform/variables.tf
Comment (martinpeck) [inline templates/workspace_services/health-services/terraform/variables.tf:51]: nit: add new-line to end of file

File: templates/workspace_services/innereye/terraform/variables.tf
Comment (martinpeck) [inline templates/workspace_services/innereye/terraform/variables.tf:12]: nit: add new-line to end of file

File: templates/workspace_services/mlflow/terraform/variables.tf
Comment (martinpeck) [inline templates/workspace_services/mlflow/terraform/variables.tf:13]: nit: add new-line to end of file

File: templates/workspace_services/mysql/terraform/variables.tf
Comment (martinpeck) [inline templates/workspace_services/mysql/terraform/variables.tf:13]: nit: add new-line to end of file

**Review by martinpeck — APPROVED** [review]
Wow. That was epic! Well done @LizaShak !

I've made a few comments, but they're mostly minor things.
Please ensure you get reviews from others (don't merge purely based upon my approval).

File: airlock_processor/shared_code/cloud.py
Comment (martinpeck) [reply airlock_processor/shared_code/cloud.py:16]: Repeated code is a repo-wide issue

If you think this specific case is fixable in this PR then great, but feel free to resolve this as "won't fix". It's fine as-is.

File: cli/tre/authentication.py
Comment (martinpeck) [reply cli/tre/authentication.py:48]: Repeated code is a repo-wide issue

If you think this specific case is fixable in this PR then great, but feel free to resolve this as "won't fix". It's fine as-is.

File: core/terraform/locals.tf
Comment (martinpeck) [reply core/terraform/locals.tf:38]: This is really the only conversation I have open that I think needs addressing.
If we can, I'd suggest some lines of comment above this regex to explain what it does and how it works, and what the expected inputs/outputs are.

**Review by tamirkamara — APPROVED** [review]
Great job!

**Pull request merged**
